File: practical1/sjf_preemptive.cpp
#include <iostream>
#include <algorithm>
#include <vector>
#include <string>
#include <iomanip>

using namespace std;


vector<string> nameOfJobs = {"P1", "P2", "P3", "P4"};
vector<int> arrivalTimes = {1, 2, 1, 4};
vector<int> burstTimes = {3, 4, 2, 4};

const int inf = 100;

class Job {
public:
    string name;
    bool served = false;
    int arrivalTime, burstTime, waitingTime, responseTime, completionTime;

    Job(string name, int arrivalTime, int burstTime){
        this->name = name;
        this->arrivalTime = arrivalTime;
        this->burstTime = burstTime;
        this->waitingTime  = 0;
        this->completionTime = 0;
        this->responseTime = 0;
    }

    void printStats(){
        cout  << setw(20) << this->name << setw(5) << this->arrivalTime << setw(5) << this->burstTime << setw(5) << this->waitingTime << setw(5) << this->completionTime << setw(5) << this->responseTime << endl;
    }
};


void insertJobsInJobQueue(vector<Job> &jobQueue)
{
    int numberOfJobs = nameOfJobs.size();

    for(int i=0; i<numberOfJobs; i++){
        string nameOfCurJob = nameOfJobs[i];
        int arrivalTimeOfCurJob = arrivalTimes[i];
        int burstTimeOfCurJob = burstTimes[i];

        Job newJob(nameOfCurJob, arrivalTimeOfCurJob, burstTimeOfCurJob);
        jobQueue.push_back(newJob);
    }
    
    
}



void insertJobsInReadyQueue(vector<Job*> &readyQueue, vector<Job> &jobQueue, int counter){

    // check for all the jobs which are arrived before counter
    // passed in original jobQueue
    for(auto &job : jobQueue){
        if(job.arrivalTime <= counter && job.served == false){
            Job* jobptr = &job;
            readyQueue.push_back(jobptr);
        }
    }

}

bool onBurstTime(Job* j1, Job* j2){
    return j1->burstTime <= j2->burstTime;
}

int main()
{
    vector<Job> jobQueue = {}, finishedJobs = {};
    insertJobsInJobQueue(jobQueue);

    // sort appropriately 
    // sort(jobQueue.begin(), jobQueue.end(), onPriority);

    
    int counter = 0;

    while(counter < inf){
        // address of all job objects
        vector<Job*> readyQueue;

        // this code is wrong
        insertJobsInReadyQueue(readyQueue, jobQueue, counter);

        // if no job has arrived, then don't do anything
        if(readyQueue.size() == 0){
            counter++;
            continue;
        } 
        sort(readyQueue.begin(), readyQueue.end(), onBurstTime);

        // // get the first job
        // Job *jobToServe = &readyQueue[0];
        Job* jobToServe = readyQueue[0]; // address of the job to serve

        // cout << jobToServe->served << endl;

        // // cout << jobToServe.name << endl;
        jobToServe->served = true;


        // counter += jobToServe->burstTime;

        jobToServe->completionTime = counter + jobToServe->burstTime;

        counter = jobToServe->completionTime;
        
    
        jobToServe->waitingTime = counter - jobToServe->arrivalTime;

        jobToServe->responseTime = jobToServe->waitingTime;

        finishedJobs.push_back(*jobToServe);

        




    }
    
    for(auto job : finishedJobs) job.printStats();


}

File: practical1/sjf_preemptive.py
class Job:
    def __init__(self, name, arrival_time, burst_time):
        self.name = name
        self.arrival_time = arrival_time
        self.burst_time = burst_time
        self.completion_time = 0
        self.waiting_time = 0
        self.response_time = 0
        self.served = False

    def print_stats(self):
        print(f'{self.name.center(10)}\t|\t{self.arrival_time}\t|\t{self.burst_time}\t|\t{self.waiting_time}\t|\t{self.completion_time}')
        

def insert_jobs_in_job_queue(job_queue, names, arrival_times, burst_times):

    number_of_jobs = len(names)

    for i in range(number_of_jobs):
        name_of_cur_job = names[i]

        arrival_time_of_current_job = arrival_times[i]

        burst_time_of_cur_job = burst_times[i]

        newJob = Job(name_of_cur_job, arrival_time_of_current_job, burst_time_of_cur_job)

        job_queue.append(newJob)

    # make a pass by referece call

def insert_jobs_in_ready_queue(ready_queue, job_queue, timer):
    for job in job_queue:
        if job.arrival_time <= timer and job.served == False:
            ready_queue.append(job)

def by_burst_time(first_job, second_job):
    return first_job.burst_time - second_job.burst_time


def serve_job(job, timer):
    job.served = True

    job.waiting_time = timer - job.arrival_time

    job.completion_time = timer + job.burst_time

    job.response_time = job.waiting_time

def print_final_information(finished_jobs, total_waiting_time):

    average_waiting_time = total_waiting_time / len(finished_jobs)
    print('----------------------------------------------------------------------------')    
    print(f'  Name\t\t\tA.T\t\tB.T\t\tW.T\t\tC.T')
    print('----------------------------------------------------------------------------')
    for job in finished_jobs:
        job.print_stats()
    print('----------------------------------------------------------------------------')

    print(f'average waiting time in the system  = {average_waiting_time}')


def main():

    names = ["P1", "P2", "P3", "P4", "P5"]
    arrival_times = [2, 5, 1, 0, 4]
    burst_times = [6, 2, 8, 3, 4]

    job_queue = []
    finished_jobs = []

    insert_jobs_in_job_queue(job_queue, names, arrival_times, burst_times)

    timer = 0
    total_waiting_time = 0
    while(timer < 100):
        ready_queue = []

        insert_jobs_in_ready_queue(ready_queue, job_queue, timer)
    
        if len(ready_queue) == 0:
            timer += 1
            continue

        ready_queue.sort(key = lambda x : x.burst_time)


        best_job_to_serve = ready_queue[0]

        serve_job(best_job_to_serve, timer) 

        timer = best_job_to_serve.completion_time

        total_waiting_time += best_job_to_serve.waiting_time

        finished_jobs.append(best_job_to_serve)


    print_final_information(finished_jobs, total_waiting_time)
    

    # make the job queue
main()